src/Orleans/Threading/CancellationTokenManager.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Runtime;
+
+namespace Orleans.Threading
+{
+    internal static class CancellationTokenManager
+    {
+        public static void WrapCancellationTokens(object[] arguments)
+        {
+            if (arguments == null) return;
+            for (var i = 0; i < arguments.Length; i++)
+            {
+                var argument = arguments[i];
+                if (argument is CancellationToken)
+                {
+                    arguments[i] = WrapCancellationToken((CancellationToken)argument);
+                }
+            }
+        }
+
+        public static CancellationTokenWrapper WrapCancellationToken(CancellationToken ct)
+        {
+            var tokenId = Guid.NewGuid();
+            return new CancellationTokenWrapper(tokenId, ct)
+            {
+                OnSerialization = () =>
+                {
+                    RegisterTokenCallbacks(ct, tokenId);
+                }
+            };
+        }
+
+        public static Task<CancellationToken> GetCancellationToken(Guid tokenId)
+        {
+            return RuntimeClient.Current.InternalGrainFactory.GetGrain<ICancellationTokenHolderGrain>(tokenId).GetCancellationToken();
+        }
+
+        private static void RegisterTokenCallbacks(CancellationToken ct, Guid tokenId)
+        {
+            if (!ct.CanBeCanceled) return;
+            ct.Register(
+               async (f) =>
```

Comment:
Just to make sure it's absolutely clear: this `async` func here is unobserved, unawaited, and is being seen by the `CancellationToken.Register` method as an `Action<object>`. If that's the intention and the implications have been carefully considered, then could we mark this with a comment discussing that? It just makes me nervous.

Nit: typically single parameter lambdas omit the superfluous parenthesis.